File: winui4k-sample-gallery-in-java/src/main/java/com/appkitbox/winui4k/sample/gallery/GalleryNavigation.java
package com.appkitbox.winui4k.sample.gallery;

import com.appkitbox.winui4k.NavigationViewBackButtonVisible;
import com.appkitbox.winui4k.Symbol;
import com.appkitbox.winui4k.WComponent;
import com.appkitbox.winui4k.WNavigationView;
import com.appkitbox.winui4k.WNavigationViewItem;
import kotlin.Unit;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiConsumer;
import java.util.function.Supplier;

/**
 * The return value of {@link #buildGalleryNavigationView}: the navigation view itself, plus
 * WNavigationViewItem" lookup used to sync selection for the back button.
 */
class GalleryNavigation {
    final WNavigationView navigationView;
    final WNavigationViewItem homeItem;
    final WNavigationViewItem settingsItem;
    final Map<String, WNavigationViewItem> itemsByPageName;

    GalleryNavigation(
            WNavigationView navigationView,
            WNavigationViewItem homeItem,
            WNavigationViewItem settingsItem,
            Map<String, WNavigationViewItem> itemsByPageName) {
        this.navigationView = navigationView;
        this.homeItem = homeItem;
        this.settingsItem = settingsItem;
        this.itemsByPageName = itemsByPageName;
    }

    /** Page name (WinUI control name) -> the function that builds its demo page. */
    static final Map<String, Supplier<WComponent>> pages = new LinkedHashMap<>();

    static {
        pages.put("AppBarButton", MenusToolbarsPages::buildAppBarButtonPage);
        pages.put("AppBarSeparator", MenusToolbarsPages::buildAppBarSeparatorPage);
        pages.put("AppBarToggleButton", MenusToolbarsPages::buildAppBarToggleButtonPage);
        pages.put("AutoSuggestBox", TextPages::buildAutoSuggestBoxPage);
        pages.put("AppNotification", ShellPages::buildAppNotificationPage);
        pages.put("BadgeNotification", ShellPages::buildBadgeNotificationPage);
        pages.put("Border", LayoutPages::buildBorderPage);
        pages.put("BreadcrumbBar", NavigationPages::buildBreadcrumbBarPage);
        pages.put("Button", BasicInputPages::buildButtonPage);
        pages.put("CalendarDatePicker", DateTimePages::buildCalendarDatePickerPage);
        pages.put("CalendarView", DateTimePages::buildCalendarViewPage);
        pages.put("Canvas", LayoutPages::buildCanvasPage);
        pages.put("CheckBox", BasicInputPages::buildCheckBoxPage);
        pages.put("ColorPicker", BasicInputPages::buildColorPickerPage);
        pages.put("ComboBox", BasicInputPages::buildComboBoxPage);
        pages.put("CommandBar", MenusToolbarsPages::buildCommandBarPage);
        pages.put("CommandBarFlyout", MenusToolbarsPages::buildCommandBarFlyoutPage);
        pages.put("ContentDialog", DialogsFlyoutsPages::buildContentDialogPage);
        pages.put("DatePicker", DateTimePages::buildDatePickerPage);
        pages.put("DropDownButton", BasicInputPages::buildDropDownButtonPage);
        pages.put("Expander", LayoutPages::buildExpanderPage);
        pages.put("Flyout", DialogsFlyoutsPages::buildFlyoutPage);
        pages.put("Grid", LayoutPages::buildGridPage);
        pages.put("HyperlinkButton", BasicInputPages::buildHyperlinkButtonPage);
        pages.put("InfoBadge", StatusInfoPages::buildInfoBadgePage);
        pages.put("InfoBar", StatusInfoPages::buildInfoBarPage);
        pages.put("JumpList", ShellPages::buildJumpListPage);
        pages.put("ItemsView", CollectionsPages::buildItemsViewPage);
        pages.put("LayoutPanel", LayoutPages::buildLayoutPanelPage);
        pages.put("ListBox", CollectionsPages::buildListBoxPage);
        pages.put("ListView", CollectionsPages::buildListViewPage);
        pages.put("MenuBar", MenusToolbarsPages::buildMenuBarPage);
        pages.put("MenuFlyout", MenusToolbarsPages::buildMenuFlyoutPage);
        pages.put("NavigationView", NavigationPages::buildNavigationViewPage);
        pages.put("NumberBox", TextPages::buildNumberBoxPage);
        pages.put("PasswordBox", TextPages::buildPasswordBoxPage);
        pages.put("PipsPager", ScrollingPages::buildPipsPagerPage);
        pages.put("Popup", DialogsFlyoutsPages::buildPopupPage);
        pages.put("ProgressBar", StatusInfoPages::buildProgressBarPage);
        pages.put("ProgressRing", StatusInfoPages::buildProgressRingPage);
        pages.put("RadioButton", BasicInputPages::buildRadioButtonPage);
        pages.put("RatingControl", BasicInputPages::buildRatingControlPage);
        pages.put("RelativePanel", LayoutPages::buildRelativePanelPage);
        pages.put("RepeatButton", BasicInputPages::buildRepeatButtonPage);
        pages.put("AnnotatedScrollBar", ScrollingPages::buildAnnotatedScrollBarPage);
        pages.put("ScrollView", ScrollingPages::buildScrollViewPage);
        pages.put("ScrollViewer", ScrollingPages::buildScrollViewerPage);
        pages.put("SelectorBar", NavigationPages::buildSelectorBarPage);
        pages.put("SemanticZoom", ScrollingPages::buildSemanticZoomPage);
        pages.put("SettingsCard", LayoutPages::buildSettingsCardPage);
        pages.put("RichEditBox", TextPages::buildRichEditBoxPage);
        pages.put("RichTextBlock", TextPages::buildRichTextBlockPage);
        pages.put("Slider", BasicInputPages::buildSliderPage);
        pages.put("SplitButton", BasicInputPages::buildSplitButtonPage);
        pages.put("SplitView", LayoutPages::buildSplitViewPage);
        pages.put("StackPanel", LayoutPages::buildStackPanelPage);
        pages.put("StandardUICommand", MenusToolbarsPages::buildStandardUICommandPage);
        pages.put("SwipeControl", MenusToolbarsPages::buildSwipeControlPage);
        pages.put("TableView", CollectionsPages::buildTableViewPage);
        pages.put("TabView", NavigationPages::buildTabViewPage);
        pages.put("TeachingTip", DialogsFlyoutsPages::buildTeachingTipPage);
        pages.put("TextBlock", TextPages::buildTextBlockPage);
        pages.put("TextBox", TextPages::buildTextBoxPage);
        pages.put("TimePicker", DateTimePages::buildTimePickerPage);
        pages.put("ToggleButton", BasicInputPages::buildToggleButtonPage);
        pages.put("ToolTip", StatusInfoPages::buildToolTipPage);
        pages.put("ToggleSplitButton", BasicInputPages::buildToggleSplitButtonPage);
        pages.put("ToggleSwitch", BasicInputPages::buildToggleSwitchPage);
        pages.put("TreeView", CollectionsPages::buildTreeViewPage);
        pages.put("WebView2", MediaPages::buildWebView2Page);
        pages.put("VariableSizedWrapGrid", LayoutPages::buildVariableSizedWrapGridPage);
        pages.put("XamlUICommand", MenusToolbarsPages::buildXamlUICommandPage);
        pages.put("AppWindow", WindowingPages::buildAppWindowPage);
        pages.put("AppWindowTitleBar", WindowingPages::buildAppWindowTitleBarPage);
        pages.put("Multiple windows", WindowingPages::buildMultipleWindowsPage);
        pages.put("SystemBackdrop", WindowingPages::buildSystemBackdropPage);
        pages.put("TitleBar", WindowingPages::buildTitleBarPage);
    }

    /** Navigation categories (matching the real WinUI 3 Gallery's grouping) -> the page names in each. */
    private static final Map<String, List<String>> categories = new LinkedHashMap<>();

    static {
        categories.put("Basic input", Arrays.asList(
                "Button",
                "CheckBox",
                "ColorPicker",
                "ComboBox",
                "DropDownButton",
                "HyperlinkButton",
                "RadioButton",
                "RatingControl",
                "RepeatButton",
                "Slider",
                "SplitButton",
                "ToggleButton",
                "ToggleSplitButton",
                "ToggleSwitch"));
        categories.put("Date & time", Arrays.asList(
                "CalendarDatePicker",
                "CalendarView",
                "DatePicker",
                "TimePicker"));
        categories.put("Collections", Arrays.asList(
                "ItemsView",
                "ListBox",
                "ListView",
                "TableView",
                "TreeView"));
        categories.put("Dialogs & flyouts", Arrays.asList(
                "ContentDialog",
                "Flyout",
                "Popup",
                "TeachingTip"));
        categories.put("Layout", Arrays.asList(
                "Border",
                "Canvas",
                "Expander",
                "Grid",
                "LayoutPanel",
                "RelativePanel",
                "SettingsCard",
                "SplitView",
                "StackPanel",
                "VariableSizedWrapGrid"));
        categories.put("Media", Arrays.asList(
                "WebView2"));
        categories.put("Menus & toolbars", Arrays.asList(
                "AppBarButton",
                "AppBarSeparator",
                "AppBarToggleButton",
                "CommandBar",
                "CommandBarFlyout",
                "MenuBar",
                "MenuFlyout",
                "SwipeControl",
                "StandardUICommand",
                "XamlUICommand"));
        categories.put("Navigation", Arrays.asList(
                "BreadcrumbBar",
                "NavigationView",
                "SelectorBar",
                "TabView"));
        categories.put("Scrolling", Arrays.asList(
                "AnnotatedScrollBar",
                "PipsPager",
                "ScrollView",
                "ScrollViewer",
                "SemanticZoom"));
        categories.put("Shell", Arrays.asList(
                "AppNotification",
                "BadgeNotification",
                "JumpList"));
        categories.put("Status & info", Arrays.asList(
                "InfoBadge",
                "InfoBar",
                "ProgressBar",
                "ProgressRing",
                "ToolTip"));
        categories.put("Text", Arrays.asList(
                "AutoSuggestBox",
                "NumberBox",
                "PasswordBox",
                "RichEditBox",
                "RichTextBlock",
                "TextBlock",
                "TextBox"));
        categories.put("Windowing", Arrays.asList(
                "AppWindow",
                "AppWindowTitleBar",
                "Multiple windows",
                "SystemBackdrop",
                "TitleBar"));
    }

    /** Category name -> the icon shown to the left of the category name in the navigation. */
    private static final Map<String, Symbol> categoryIcons = new LinkedHashMap<>();

    static {
        categoryIcons.put("Basic input", Symbol.KEYBOARD);
        categoryIcons.put("Date & time", Symbol.CALENDAR);
        categoryIcons.put("Collections", Symbol.LIST);
        categoryIcons.put("Dialogs & flyouts", Symbol.COMMENT);
        categoryIcons.put("Layout", Symbol.VIEW_ALL);
        categoryIcons.put("Media", Symbol.GLOBE);
        categoryIcons.put("Menus & toolbars", Symbol.SAVE);
        categoryIcons.put("Navigation", Symbol.GLOBAL_NAVIGATION_BUTTON);
        categoryIcons.put("Scrolling", Symbol.ZOOM);
        categoryIcons.put("Shell", Symbol.MESSAGE);
        categoryIcons.put("Status & info", Symbol.IMPORTANT);
        categoryIcons.put("Text", Symbol.FONT);
        categoryIcons.put("Windowing", Symbol.NEW_WINDOW);
    }

    /**
     * The left-hand navigation. Puts Home first, then lines up page items under each category's
     * lists the page items. Selecting Home calls {@code onHome} and selecting a page calls
     * page passes the selected page's name and builder to [onSelect].
     */
    static GalleryNavigation buildGalleryNavigationView(
            Runnable onHome,
            Runnable onSettings,
            BiConsumer<String, Supplier<WComponent>> onSelect) {
        WNavigationView navigationView = new WNavigationView();
        // Puts Settings (a gear) at the bottom-left of the pane, same as the official Gallery. The built-in
        // settings item is not used because its label is localized to the OS language, so an English footer item is added instead
        navigationView.setSettingsVisible(false);
        navigationView.setBackButtonVisible(NavigationViewBackButtonVisible.COLLAPSED);
        navigationView.setOpenPaneLength(260.0);

        // Put Home first, matching the real Gallery
        WNavigationViewItem homeItem = new WNavigationViewItem("Home", Symbol.HOME);
        navigationView.addItem(homeItem);

        Map<String, WNavigationViewItem> itemsByPageName = new LinkedHashMap<>();
        for (Map.Entry<String, List<String>> entry : categories.entrySet()) {
            String category = entry.getKey();
            // Categories aren't selectable (SelectsOnInvoked=false); they only toggle their children open/closed
            WNavigationViewItem categoryItem = new WNavigationViewItem(category, categoryIcons.get(category));
            categoryItem.setSelectsOnInvoked(false);
            categoryItem.setExpanded(false); // keep every category collapsed at startup
            for (String name : entry.getValue()) {
                WNavigationViewItem pageItem = new WNavigationViewItem(name, null);
                categoryItem.addItem(pageItem);
                itemsByPageName.put(name, pageItem);
            }
            navigationView.addItem(categoryItem);
        }

        WNavigationViewItem settingsItem = new WNavigationViewItem(SettingsPage.SETTINGS_PAGE_NAME, Symbol.SETTING);
        navigationView.addFooterItem(settingsItem);

        navigationView.addSelectionListener(item -> {
            if (item == null) {
                return Unit.INSTANCE;
            }
            if (item == homeItem) {
                onHome.run();
            } else if (item == settingsItem) {
                onSettings.run();
            } else {
                Supplier<WComponent> buildPage = pages.get(item.getText());
                if (buildPage != null) {
                    onSelect.accept(item.getText(), buildPage);
                }
            }
            return Unit.INSTANCE;
        });
        return new GalleryNavigation(navigationView, homeItem, settingsItem, itemsByPageName);
    }
}

File: winui4k-sample-gallery/src/main/kotlin/com/appkitbox/winui4k/sample/gallery/GalleryNavigation.kt
package com.appkitbox.winui4k.sample.gallery

import com.appkitbox.winui4k.NavigationViewBackButtonVisible
import com.appkitbox.winui4k.Symbol
import com.appkitbox.winui4k.WComponent
import com.appkitbox.winui4k.WNavigationView
import com.appkitbox.winui4k.WNavigationViewItem

/** Page name (WinUI control name) -> the function that builds its demo page. */
internal val pages: Map<String, () -> WComponent> = linkedMapOf(
    "AppBarButton" to ::buildAppBarButtonPage,
    "AppBarSeparator" to ::buildAppBarSeparatorPage,
    "AppBarToggleButton" to ::buildAppBarToggleButtonPage,
    "AutoSuggestBox" to ::buildAutoSuggestBoxPage,
    "AppNotification" to ::buildAppNotificationPage,
    "BadgeNotification" to ::buildBadgeNotificationPage,
    "Border" to ::buildBorderPage,
    "Button" to ::buildButtonPage,
    "CalendarDatePicker" to ::buildCalendarDatePickerPage,
    "CalendarView" to ::buildCalendarViewPage,
    "Canvas" to ::buildCanvasPage,
    "CheckBox" to ::buildCheckBoxPage,
    "ColorPicker" to ::buildColorPickerPage,
    "ComboBox" to ::buildComboBoxPage,
    "CommandBar" to ::buildCommandBarPage,
    "CommandBarFlyout" to ::buildCommandBarFlyoutPage,
    "ContentDialog" to ::buildContentDialogPage,
    "DatePicker" to ::buildDatePickerPage,
    "DropDownButton" to ::buildDropDownButtonPage,
    "Expander" to ::buildExpanderPage,
    "Flyout" to ::buildFlyoutPage,
    "Grid" to ::buildGridPage,
    "HyperlinkButton" to ::buildHyperlinkButtonPage,
    "InfoBadge" to ::buildInfoBadgePage,
    "InfoBar" to ::buildInfoBarPage,
    "JumpList" to ::buildJumpListPage,
    "ListBox" to ::buildListBoxPage,
    "ListView" to ::buildListViewPage,
    "MenuBar" to ::buildMenuBarPage,
    "MenuFlyout" to ::buildMenuFlyoutPage,
    "NavigationView" to ::buildNavigationViewPage,
    "NumberBox" to ::buildNumberBoxPage,
    "PasswordBox" to ::buildPasswordBoxPage,
    "PipsPager" to ::buildPipsPagerPage,
    "Popup" to ::buildPopupPage,
    "ProgressBar" to ::buildProgressBarPage,
    "ProgressRing" to ::buildProgressRingPage,
    "RadioButton" to ::buildRadioButtonPage,
    "RatingControl" to ::buildRatingControlPage,
    "RelativePanel" to ::buildRelativePanelPage,
    "RepeatButton" to ::buildRepeatButtonPage,
    "AnnotatedScrollBar" to ::buildAnnotatedScrollBarPage,
    "ScrollView" to ::buildScrollViewPage,
    "ScrollViewer" to ::buildScrollViewerPage,
    "SemanticZoom" to ::buildSemanticZoomPage,
    "RichEditBox" to ::buildRichEditBoxPage,
    "RichTextBlock" to ::buildRichTextBlockPage,
    "Slider" to ::buildSliderPage,
    "SplitButton" to ::buildSplitButtonPage,
    "SplitView" to ::buildSplitViewPage,
    "StackPanel" to ::buildStackPanelPage,
    "StandardUICommand" to ::buildStandardUICommandPage,
    "SwipeControl" to ::buildSwipeControlPage,
    "TableView" to ::buildTableViewPage,
    "TeachingTip" to ::buildTeachingTipPage,
    "TextBlock" to ::buildTextBlockPage,
    "TextBox" to ::buildTextBoxPage,
    "TimePicker" to ::buildTimePickerPage,
    "ToggleButton" to ::buildToggleButtonPage,
    "ToolTip" to ::buildToolTipPage,
    "ToggleSplitButton" to ::buildToggleSplitButtonPage,
    "ToggleSwitch" to ::buildToggleSwitchPage,
    "TreeView" to ::buildTreeViewPage,
    "VariableSizedWrapGrid" to ::buildVariableSizedWrapGridPage,
    "XamlUICommand" to ::buildXamlUICommandPage,
    "AppWindow" to ::buildAppWindowPage,
    "AppWindowTitleBar" to ::buildAppWindowTitleBarPage,
    "Multiple windows" to ::buildMultipleWindowsPage,
    "SystemBackdrop" to ::buildSystemBackdropPage,
    "TitleBar" to ::buildTitleBarPage,
)

/** Navigation categories (matching the real WinUI 3 Gallery's grouping) -> the page names in each. */
private val categories: Map<String, List<String>> = linkedMapOf(
    "Basic input" to listOf(
        "Button",
        "CheckBox",
        "ColorPicker",
        "ComboBox",
        "DropDownButton",
        "HyperlinkButton",
        "RadioButton",
        "RatingControl",
        "RepeatButton",
        "Slider",
        "SplitButton",
        "ToggleButton",
        "ToggleSplitButton",
        "ToggleSwitch",
    ),
    "Date & time" to listOf(
        "CalendarDatePicker",
        "CalendarView",
        "DatePicker",
        "TimePicker",
    ),
    "Collections" to listOf(
        "ListBox",
        "ListView",
        "TableView",
        "TreeView",
    ),
    "Dialogs & flyouts" to listOf(
        "ContentDialog",
        "Flyout",
        "Popup",
        "TeachingTip",
    ),
    "Layout" to listOf(
        "Border",
        "Canvas",
        "Expander",
        "Grid",
        "RelativePanel",
        "SplitView",
        "StackPanel",
        "VariableSizedWrapGrid",
    ),
    "Menus & toolbars" to listOf(
        "AppBarButton",
        "AppBarSeparator",
        "AppBarToggleButton",
        "CommandBar",
        "CommandBarFlyout",
        "MenuBar",
        "MenuFlyout",
        "SwipeControl",
        "StandardUICommand",
        "XamlUICommand",
    ),
    "Navigation" to listOf(
        "NavigationView",
    ),
    "Scrolling" to listOf(
        "AnnotatedScrollBar",
        "PipsPager",
        "ScrollView",
        "ScrollViewer",
        "SemanticZoom",
    ),
    "Shell" to listOf(
        "AppNotification",
        "BadgeNotification",
        "JumpList",
    ),
    "Status & info" to listOf(
        "InfoBadge",
        "InfoBar",
        "ProgressBar",
        "ProgressRing",
        "ToolTip",
    ),
    "Text" to listOf(
        "AutoSuggestBox",
        "NumberBox",
        "PasswordBox",
        "RichEditBox",
        "RichTextBlock",
        "TextBlock",
        "TextBox",
    ),
    "Windowing" to listOf(
        "AppWindow",
        "AppWindowTitleBar",
        "Multiple windows",
        "SystemBackdrop",
        "TitleBar",
    ),
)

/** Category name -> the icon shown to the left of the category name in the navigation. */
private val categoryIcons: Map<String, Symbol> = mapOf(
    "Basic input" to Symbol.KEYBOARD,
    "Date & time" to Symbol.CALENDAR,
    "Collections" to Symbol.LIST,
    "Dialogs & flyouts" to Symbol.COMMENT,
    "Layout" to Symbol.VIEW_ALL,
    "Menus & toolbars" to Symbol.SAVE,
    "Navigation" to Symbol.GLOBAL_NAVIGATION_BUTTON,
    "Scrolling" to Symbol.ZOOM,
    "Shell" to Symbol.MESSAGE,
    "Status & info" to Symbol.IMPORTANT,
    "Text" to Symbol.FONT,
    "Windowing" to Symbol.NEW_WINDOW,
)

/**
 * [buildGalleryNavigationView]'s return value: the navigation itself, plus a "page name ->
 * WNavigationViewItem" lookup used to sync selection for the back button.
 */
internal class GalleryNavigation(
    val navigationView: WNavigationView,
    val homeItem: WNavigationViewItem,
    val settingsItem: WNavigationViewItem,
    val itemsByPageName: Map<String, WNavigationViewItem>,
)

/**
 * The left-hand navigation. Puts Home first, then lines up page items under each category's
 * parent item (with an icon) below it. Selecting Home passes control to [onHome]; selecting a
 * page passes the selected page's name and builder to [onSelect].
 */
internal fun buildGalleryNavigationView(
    onHome: () -> Unit,
    onSettings: () -> Unit,
    onSelect: (String, () -> WComponent) -> Unit,
): GalleryNavigation {
    val navigationView = WNavigationView()
    // Show Settings (gear) at the pane's bottom-left, matching the real Gallery. Not using the
    // built-in Settings entry since its label gets localized to the OS's language ("Settings" ->
    // something else), so a self-built English footer item is used instead
    navigationView.isSettingsVisible = false
    navigationView.isBackButtonVisible = NavigationViewBackButtonVisible.COLLAPSED
    navigationView.openPaneLength = 260.0

    // Put Home first, matching the real Gallery
    val homeItem = WNavigationViewItem("Home", Symbol.HOME)
    navigationView.addItem(homeItem)

    val itemsByPageName = mutableMapOf<String, WNavigationViewItem>()
    for ((category, names) in categories) {
        // Categories aren't selectable (SelectsOnInvoked=false); they only toggle their children open/closed
        val categoryItem = WNavigationViewItem(category, categoryIcons[category])
        categoryItem.selectsOnInvoked = false
        categoryItem.isExpanded = false // start collapsed on launch
        for (name in names) {
            val pageItem = WNavigationViewItem(name)
            categoryItem.addItem(pageItem)
            itemsByPageName[name] = pageItem
        }
        navigationView.addItem(categoryItem)
    }

    val settingsItem = WNavigationViewItem(SETTINGS_PAGE_NAME, Symbol.SETTING)
    navigationView.addFooterItem(settingsItem)

    navigationView.addSelectionListener { item ->
        when {
            item == null -> return@addSelectionListener
            item == homeItem -> onHome()
            item == settingsItem -> onSettings()
            else -> pages[item.text]?.let { buildPage -> onSelect(item.text, buildPage) }
        }
    }
    return GalleryNavigation(navigationView, homeItem, settingsItem, itemsByPageName)
}
